proposals/0000-code-texp.rst

```diff
@@ -0,0 +1,167 @@
+Make Q (TExp a) into a newtype
+==============
+
+.. proposal-number:: Leave blank. This will be filled in when the proposal is
+                     accepted.
+.. trac-ticket:: #16177
+.. implemented:: Leave blank. This will be filled in with the first GHC version which
+                 implements the described feature.
+.. highlight:: haskell
+.. header:: This proposal is `discussed at this pull request <https://github.com/ghc-proposals/ghc-proposals/pull/0>`_.
+            **After creating the pull request, edit this file again, update the
+            number in the link, and delete this bold sentence.**
+.. sectnum::
+.. contents::
+
+I propose to modify the typed template Haskell API to make the ``Code`` type
+more abstract. In particular we introduce a new data type called ``Code`` which
+is abstract such that ``Code m a`` represents an expression of type ``a`` produced
+in a monadic context ``m``. The reader should note that this proposal builds on
+top of the `overloaded quotations proposal<https://github.com/ghc-proposals/ghc-proposals/pull/246>`_ which was
+accepted before this proposal.
+
+
+Motivation
+------------
+
+There are three problems with the current API:
+
+1. It is hard to properly write instances for ``Quote m => m (TExp a)`` as the type is the composition
+   of two type constructors. Doing so in your program involves making your own newtype and
+   doing a lot of wrapping/unwrapping.
+
+   For example, if I want to create a language which I can either run immediately or
+   generate code from I could write the following with the new API. ::
+
+      class Lang r where
+        _int :: Int -> r Int
+        _if  :: r Bool -> r a -> r a -> r a
+
+      instance Lang Identity where
+        _int = Identity
+        _if (Identity b) (Identity t) (Identity f) = Identity (if b then t else f)
+
+      instance Lang (Code Q) where
+        _int = liftTyped
+        _if cb ct cf = [|| if $$cb then $$ct else $$cf ||]
+
+2. When doing code generation it is common to want to store code fragments in
+   a map. When doing typed code generation, these code fragments contain a
+   type index so it is desirable to store them in one of the parameterised
+   map data types such as ``DMap`` from ``dependent-map`` or ``MapF`` from
+   ``parameterized-utils``.
+
+   ::
+
+      compiler :: Env -> AST a -> Code Q a
+
+      data AST a where ...
+      data Ident a = ...
+
+      type Env = MapF Ident (Code Q)
+
+      newtype Code m a = Code (m (TExp a))
+
+
+   In this example, the ``MapF`` maps an ``Ident String`` directly to a ``Code Q String``.
+   Using one of these map types currently requires creating your own newtype and constantly
+   wrapping every quotation and unwrapping it when using a splice. Achievable, but
+   it creates even more syntactic noise than normal metaprogramming.
+
+
+3. ``Quote m => m (TExp a)`` is ugly to read and write. This is a weak reason but one everyone
```

Comment:
Does `Quote m => m (TExp a)` have worse readability than `Quote m => Code m a`? Maybe a little bit, but not as much as to call it "ugly".